### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by martinRenou**
Fix https://github.com/jupytercad/jupytercad/issues/343

https://github.com/jupytercad/jupytercad/assets/21197331/15a22bb3-46f9-40ca-a960-815b48069a93

### CODE REVIEW

**Comment by trungleduc** [comment]
Thanks @martinRenou for working on it. I just tested with some corner cases:

- Adding shape or editing shape from the right panel does not go through the dry-run check, should we do it?

https://github.com/jupytercad/jupytercad/assets/4451292/f7d51cba-98f0-4889-b003-2f8af6804189

- I artificially make the `loadFlle` action much slower so that I can add another operator while the first one is still processing. But then the first operator is discarded, we may need to lock the operators while the worker is busy

https://github.com/jupytercad/jupytercad/assets/4451292/e93699ee-5288-436c-8bb3-c6788062bbd3

- When there are 2 users, the first user adds an operator but since the shared doc is not updated, the second user might add another operator with the same name

https://github.com/jupytercad/jupytercad/assets/4451292/17cb5797-a1a6-435c-b262-962d2968b846

- Even if the second user can avoid the naming collision, the outcome is not correct since their worker does not know about the upcoming operator of the first user


https://github.com/jupytercad/jupytercad/assets/4451292/f99fe843-25ca-46dc-96e5-890ff52289b9

I think we might need to improve the loading indicator and locking mechanism to avoid these cases

**Comment by martinRenou** [comment]
> Adding shape or editing shape from the right panel does not go through the dry-run check, should we do it?

Yes, definitely. Also we should probably do it on the first load of the file, and report malformed a file if that's the case.

> I artificially make the loadFlle action much slower so that I can add another operator while the first one is still processing. But then the first operator is discarded, we may need to lock the operators while the worker is busy

Right! We should fix it indeed. It's unrelated to this PR though, so we should probably turn this into an issue and look into implementing a fix.

> When there are 2 users, the first user adds an operator but since the shared doc is not updated, the second user might add another operator with the same name

I believe name conflicts should not be a thing. We can fix it by fixing https://github.com/jupytercad/jupytercad/issues/200. 

> Even if the second user can avoid the naming collision, the outcome is not correct since their worker does not know about the upcoming operator of the first user

I guess this would be fixed by implementing a proper lock?

**Comment by trungleduc** [comment]
yeah, having a proper operator lock definitely helps. But I also feel that having a single promise delegate for all operators might be a bottleneck. Using a dict of promise delegates and sending the delegate id back and forth between the worker and the main view model can work?

**Comment by martinRenou** [comment]
Oh I see. I'll have a look at that.

**Review by trungleduc — APPROVED** [review]
Thanks 🚀

**Comment by martinRenou** [comment]
Wait before merging I'll check to use the dry run in the object properties panel too :)

**Comment by martinRenou** [comment]
@trungleduc Ready for review!

File: packages/base/src/panelview/objectproperties.tsx [hunk 1]
```diff
@@ -97,18 +101,51 @@ class ObjectPropertiesReact extends React.Component<IProps, IStates> {
...
   ) {
     if (!this.state.jcadObject || !objectName) {
       return;
     }
 
-    const model = this.props.cpModel.jcadModel?.sharedModel;
-    const obj = model?.getObjectByName(objectName);
-    if (model && obj) {
-      model.updateObjectByName(objectName, 'parameters', {
+    const currentViewModel = this.props.tracker
```
Comment (trungleduc) [inline packages/base/src/panelview/objectproperties.tsx:112]:
```suggestion
    const currentWidget= this.props.tracker
```

File: packages/base/src/panelview/objectproperties.tsx [hunk 2]
```diff
@@ -97,18 +101,51 @@ class ObjectPropertiesReact extends React.Component<IProps, IStates> {
...
+      .currentWidget as JupyterCadWidget | null;
+    if (!currentViewModel) {
```
Comment (trungleduc) [inline packages/base/src/panelview/objectproperties.tsx:114]:
```suggestion
    if (!currentWidget) {
```

File: packages/base/src/panelview/objectproperties.tsx [hunk 3]
```diff
@@ -97,18 +101,51 @@ class ObjectPropertiesReact extends React.Component<IProps, IStates> {
...
+        object.parameters = {
+          ...object.parameters,
+          ...properties
+        };
+      }
+    }
+
+    // Try a dry run
+    const dryRunResult =
+      await currentViewModel.content.currentViewModel.dryRun(updatedContent);
```
Comment (trungleduc) [inline packages/base/src/panelview/objectproperties.tsx:136]:
```suggestion
      await currentWidget.content.currentViewModel.dryRun(updatedContent);
```

File: packages/base/src/panelview/objectproperties.tsx [hunk 4]
```diff
@@ -97,18 +101,51 @@ class ObjectPropertiesReact extends React.Component<IProps, IStates> {
...
+    if (dryRunResult.status === 'error') {
+      showErrorMessage(
+        'Failed to update the shape',
+        'The tool was unable to update the desired shape due to invalid parameter values. The values you entered may not be compatible with the dimensions of your piece.'
+      );
+      return;
+    }
+
+    // Dry run was successful, really apply the update now
```
Comment (trungleduc) [inline packages/base/src/panelview/objectproperties.tsx:145]:
```suggestion
    // Dry run was successful, ready to apply the update now
```

**Comment by SylvainCorlay** [comment]
🎉  that is pretty great!

**Pull request merged**